File: app/src/main/java/com/zest/android/util/NetworkStateReceiver.java
package com.zest.android.util;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.util.Log;

import java.util.HashSet;
import java.util.Set;

public class NetworkStateReceiver extends BroadcastReceiver {


    private static final String TAG = NetworkStateReceiver.class.getSimpleName();
    protected Set<OnNetworkStateReceiverListener> listeners;
    protected Boolean connected;

    public NetworkStateReceiver() {
        listeners = new HashSet<>();
        connected = null;
    }

    public void onReceive(Context context, Intent intent) {
        ConnectivityManager connectivityManager =
                (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo activeNetInfo = connectivityManager.getActiveNetworkInfo();
        connected = activeNetInfo != null && activeNetInfo.isConnectedOrConnecting();
        notifyStateToAll();
    }

    private void notifyStateToAll() {
        Log.d(TAG, "notifyStateToAll() called");
        for (OnNetworkStateReceiverListener listener : listeners)
            notifyState(listener);
    }

    private void notifyState(OnNetworkStateReceiverListener listener) {
        if (connected == null || listener == null) return;

        if (connected)
            listener.networkAvailable();
        else
            listener.networkUnavailable();
    }

    public void addListener(OnNetworkStateReceiverListener listener) {
        listeners.add(listener);
        //notifyState(listener);
    }

    public void removeListener(OnNetworkStateReceiverListener listener) {
        listeners.remove(listener);
    }

    public interface OnNetworkStateReceiverListener {

        void networkAvailable();

        void networkUnavailable();
    }
}

File: app/src/main/java/com/zest/android/util/NetworkStateReceiver.kt
package com.zest.android.util

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.net.ConnectivityManager
import android.util.Log
import java.util.*

class NetworkStateReceiver : BroadcastReceiver() {

    protected var listeners: MutableSet<OnNetworkStateReceiverListener>
    protected var connected: Boolean? = null

    init {
        listeners = HashSet()
        connected = null
    }

    override fun onReceive(context: Context, intent: Intent) {
        val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val activeNetInfo = connectivityManager.activeNetworkInfo
        connected = activeNetInfo != null && activeNetInfo.isConnectedOrConnecting
        notifyStateToAll()
    }

    private fun notifyStateToAll() {
        Log.d(TAG, "notifyStateToAll() called")
        for (listener in listeners)
            notifyState(listener)
    }

    private fun notifyState(listener: OnNetworkStateReceiverListener?) {
        if (connected == null || listener == null) return

        if (connected!!)
            listener.networkAvailable()
        else
            listener.networkUnavailable()
    }

    fun addListener(listener: OnNetworkStateReceiverListener) {
        listeners.add(listener)
        //notifyState(listener);
    }

    fun removeListener(listener: OnNetworkStateReceiverListener) {
        listeners.remove(listener)
    }

    interface OnNetworkStateReceiverListener {

        fun networkAvailable()

        fun networkUnavailable()
    }

    companion object {

        private val TAG = NetworkStateReceiver::class.java!!.getSimpleName()
    }
}
